packages/firestore/src/api/database.ts

```diff
@@ -1031,11 +1045,24 @@ export class DocumentSnapshot implements firestore.DocumentSnapshot {
   }
 
   get metadata(): firestore.SnapshotMetadata {
-    return {
-      hasPendingWrites:
-        this._document !== null && this._document.hasLocalMutations,
-      fromCache: this._fromCache
-    };
+    return new SnapshotMetadata(
+      this._document !== null && this._document.hasLocalMutations,
+      this._fromCache
+    );
+  }
+
+  public isEqual(other: firestore.DocumentSnapshot): boolean {
+    if (!(other instanceof DocumentSnapshot)) {
+      throw invalidClassError('isEqual', 'DocumentSnapshot', 1, other);
+    }
+    return (
+      this.firestore === other.firestore &&
+      this._key.isEqual(other._key) &&
+      (this._document === null
+        ? other._document === null
+        : this._document.isEqual(other._document)) &&
+      this._fromCache === other._fromCache
```

Comment:
Nit: I would move the _fromCache comparison up since it's cheaper.

Reply:
done